### PR TITLE
Advanced Qualification Features for Study Programme

### PR DESCRIPTION
**Pull request by dkloepfer**
Contains these features for the [Advanced Qualification Quest](https://docu.ilias.de/goto.php?target=wiki_1357_Advanced_Qualification_Features):

* [Study Programme: Add link to course that was passed](https://docu.ilias.de/goto_docu_wiki_wpage_5768_1357.html)
* [Study Programme: Automatically add courses from category](https://docu.ilias.de/goto_docu_wiki_wpage_5787_1357.html)
* [Study Programme: Automatically add users by role, membership or organizational unit](https://docu.ilias.de/goto_docu_wiki_wpage_5772_1357.html)
* [Study Programme: Certificate](https://docu.ilias.de/goto_docu_wiki_wpage_5626_1357.html)
* [Study Programme: Filters and More Columns for Member List](https://docu.ilias.de/goto_docu_wiki_wpage_5822_1357.html)
* [Study Programme: Link-Object](https://docu.ilias.de/goto_docu_wiki_wpage_5769_1357.html)
* [Study Programme: Mail to Members](https://docu.ilias.de/goto_docu_wiki_wpage_5867_1357.html)
* [Study Programme: Multilingualism](https://docu.ilias.de/goto_docu_wiki_wpage_5655_1357.html)
* [Study Programme: Processing time](https://docu.ilias.de/goto_docu_wiki_wpage_5660_1357.html)
* [Study Programme: Use positions for permissions on users](https://docu.ilias.de/goto_docu_wiki_wpage_5816_1357.html)
* [Study programme: Validity of qualification](https://docu.ilias.de/goto_docu_wiki_wpage_5625_1357.html)
* [Study Programme: Automatic Info-Mails for Members](https://docu.ilias.de/goto_docu_wiki_wpage_5865_1357.html)
* [Study Programme: Mass-Operations with Members](https://docu.ilias.de/goto_docu_wiki_wpage_5868_1357.html)
* [User Search: Add Search for Organisational Unit as search option](https://docu.ilias.de/goto_docu_wiki_wpage_5771_1357.html)

### CODE REVIEW

**Comment by klees** [comment]
Hi @dkloepfer!

Please rebase this upon the current trunk (and then force-push) to get rid of those merge-commits. Please also Prefix the commits with "Prg", if applicable, or other component shortnames. Please try to clean the commit history by squashing stuff.

Thx!

**Comment by dkloepfer** [comment]
Will be reopened after rebase.

**Comment by klees** [comment]
Dear @mstuder,

this contains marginal changes in the OrgUnits:

**Comment by klees** [comment]
Dear @mstuder,

this contains some marginal changes in the OrgUnits:

* [add context for prg](https://github.com/ILIAS-eLearning/ILIAS/pull/2235/files#diff-223ea256077180a69fd4d33974f1128c)
* [add operations for prg](https://github.com/ILIAS-eLearning/ILIAS/pull/2235/files#diff-c8db9b2c542e4db3a48dc019a6eb5f7c)
* [raise events when positions are set](https://github.com/ILIAS-eLearning/ILIAS/pull/2235/files#diff-7a82b855c0a0101b389bd060b0329057)

Please approve the changes or tell us what to do differently.

Thx!

**Comment by mstuder** [comment]
Hi @klees 

The changes concerning the organizational units are OK for me.

Best regards
Martin

**Comment by klees** [comment]
Hi @smeyer-ilias,

this contains changes in your components:
* [add "Organisational Unit" as option in user search](https://github.com/ILIAS-eLearning/ILIAS/pull/2235/files#diff-17ae9d33dbef5b18ccccd4d89c2c4368)
* various marginal changes in Services/Repository, Services/Object and  Services/ContainerReference to make reference objects for Study Programmes happen

Please approve the changes or tell us what to do differently.

Thx!

**Comment by klees** [comment]
Hi @mjansenDatabay,

this contains various marginal changes in Services/Certificate to make certificates for Study Programmes happen.

Please approve the changes or tell us what to do differently.

Thx!

File: Services/Certificate/classes/Form/Repository/class.ilCertificateSettingsScormFormRepository.php [hunk 1]
```diff
@@ -85,7 +85,7 @@ public function __construct(
      * @throws ilException
      * @throws ilWACException
      */
-    public function createForm(ilCertificateGUI $certificateGUI)
+    public function createForm(ilCertificateGUI $certificateGUI, ilCertificate $certificateObject)
```
Comment (mjansenDatabay) [inline Services/Certificate/classes/Form/Repository/class.ilCertificateSettingsScormFormRepository.php]: Do you really need this?

**Review by mjansenDatabay — APPROVED** [review]
LGMT @klees .

FYI: I added some missing information to the README in `Services/Certificate` 1-2 weeks ago. I hope the pain for adding a new consumer was not too big without this documentation. Sorry for this.

File: Services/Certificate/classes/class.ilCertificateAppEventListener.php [hunk 2]
```diff
@@ -330,4 +340,33 @@ private function processEntry($type, $objectId, int $userId, ilCertificateTempla
...
     }
+
+    private function handleCompletedStudyProgramme()
+    {
+        $settings = new ilSetting('certificate');
+        $objectId = $this->parameters['prg_id'] ?? 0;
+        $userId = $this->parameters['usr_id'] ?? 0;
+        try {
+            $template = $this->templateRepository->fetchCurrentlyActiveCertificate($objectId);
+            if (true === $template->isCurrentlyActive()) {
```
Comment (mjansenDatabay) [inline Services/Certificate/classes/class.ilCertificateAppEventListener.php:351]: @klees Maybe `\ilCertificateAppEventListener::processEntry` could be called here instead?

**Comment by smeyer-ilias** [comment]
No objections from my side.

**Comment by klees** [comment]
Hi @alex40724,

could you have a look here regarding the integration in the dashboard in Services/Dashboard/classes/class.ilDashboardGUI.php?

Thx!

**Review by alex40724 — CHANGES_REQUESTED** [review]
Dashboard related changes seem to be ok, but why is there a new class Modules/Category/classes/class.ilCategoryExplorerGUI.php ?

I am out of office for today...

**Comment by klees** [comment]
Hi @alex40724,

sry, did not notice that myself until just now, but you were faster.

For [this feature](https://docu.ilias.de/goto_docu_wiki_wpage_5787_1357.html) a category needs to be selected. But I think this should be doable without a new `ExplorerGUI`. Will have a look and inform you tomorrow.

Have a nice evening!

**Comment by klees** [comment]
Hi @alex40724,

I removed that stuff by replacing it with `ilRepositorySelector2InputGUI`. Sorry again.

There are also some minor changes in `Services/ContainerReference` and `Services/Repository`to accomodate the new `StudyProgrammeReference`.

Best regards!

**Review by alex40724 — APPROVED** [review]
Looks good. Thanks.

**Review by smeyer-ilias — APPROVED** [review]
LGTM

**Pull request merged**